log/ratelimit_test.go

```diff
@@ -0,0 +1,176 @@
+
+	r.Errorln("error 2 will be discarded")
+	assert.Equal(t, 4, c.count)
+
+	r.Warnln("warning 2 will be discarded")
+	assert.Equal(t, 4, c.count)
+
+	require.NoError(t, testutil.GatherAndCompare(reg, strings.NewReader(`
+		# HELP rate_limit_logger_discarded_log_lines_total Total number of discarded log lines per level.
+		# TYPE rate_limit_logger_discarded_log_lines_total counter
+        rate_limit_logger_discarded_log_lines_total{level="debug"} 1
+        rate_limit_logger_discarded_log_lines_total{level="error"} 1
+        rate_limit_logger_discarded_log_lines_total{level="warning"} 2
+	`)))
+}
+
+func TestRateLimitedLoggerWithFields(t *testing.T) {
+	buf := bytes.NewBuffer(nil)
+	c := newCounterLogger(buf)
+	reg := prometheus.NewPedanticRegistry()
+	logger := NewRateLimitedLogger(c, 0.0001, 1, reg)
+	loggerWithFields := logger.WithField("key", "value")
+
+	loggerWithFields.Errorln("Error will be logged")
+	assert.Equal(t, 1, c.count)
+	c.verify(t, []string{"key", "value", "error", "Error will be logged"})
+
+	logger.Infoln("Info will not be logged")
+	loggerWithFields.Debugln("Debug will not be logged")
+	loggerWithFields.Warnln("Warning will not be logged")
+	assert.Equal(t, 1, c.count)
+
+	require.NoError(t, testutil.GatherAndCompare(reg, strings.NewReader(`
+		# HELP rate_limit_logger_discarded_log_lines_total Total number of discarded log lines per level.
+		# TYPE rate_limit_logger_discarded_log_lines_total counter
+        rate_limit_logger_discarded_log_lines_total{level="info"} 1
+        rate_limit_logger_discarded_log_lines_total{level="debug"} 1
+        rate_limit_logger_discarded_log_lines_total{level="warning"} 1
+	`)))
+}
+
+type counterLogger struct {
+	logger Interface
+	buf    *bytes.Buffer
+	count  int
+}
+
+func (c *counterLogger) Debugf(format string, args ...interface{}) {
+	c.logger.Debugf(format, args...)
+	c.count++
+}
+
+func (c *counterLogger) Debugln(args ...interface{}) {
+	c.logger.Debugln(args...)
+	c.count++
+}
+
+func (c *counterLogger) Infof(format string, args ...interface{}) {
+	c.logger.Infof(format, args...)
+	c.count++
+}
+
+func (c *counterLogger) Infoln(args ...interface{}) {
+	c.logger.Infoln(args...)
+	c.count++
+}
+
+func (c *counterLogger) Warnf(format string, args ...interface{}) {
+	c.logger.Warnf(format, args...)
+	c.count++
+}
+
+func (c *counterLogger) Warnln(args ...interface{}) {
+	c.logger.Warnln(args...)
+	c.count++
+}
+
+func (c *counterLogger) Errorf(format string, args ...interface{}) {
+	c.logger.Errorf(format, args...)
+	c.count++
+}
+
+func (c *counterLogger) Errorln(args ...interface{}) {
+	c.logger.Errorln(args...)
+	c.count++
+}
+
+func (c *counterLogger) WithField(key string, value interface{}) Interface {
+	c.logger = c.logger.WithField(key, value)
+	return c
+}
+
+func (c *counterLogger) WithFields(fields Fields) Interface {
+	c.logger = c.logger.WithFields(fields)
+	return c
+}
+
+func (c *counterLogger) verify(t *testing.T, logContains []string) {
```

Comment:
[nit] I think this renaming makes tests easier to read.

```suggestion
func (c *counterLogger) assertContains(t *testing.T, logContains []string) {
```